### PR TITLE
Print external WMTS (both KVP and RESTful)

### PR DESCRIPTION
**Pull request by procrastinatio**
Let's unleash the *Pandora*'s box.

**External WMTS (RESTful)**

*GetCapbilities*  https://mf-chsdi3.prod.bgdi.ch/1.0.0/WMTSCapabilities.xml

[Demo](https://mf-geoadmin3.int.bgdi.ch/mom_kvp_print/index.html?lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=WMTS%7C%7Cch.swisstopo.geologie-geologischer_atlas%7C%7Chttps:%2F%2Fmf-chsdi3.prod.bgdi.ch%2F1.0.0%2FWMTSCapabilities.xml&X=176312.50&Y=573610.00&zoom=8)

**External WMTS (RESTful, ZH)**

*GetCapbilities*  http://www.gis.stadt-zuerich.ch/wmts/wmts-zh-stzh-ogd.xml

[Demo](https://mf-geoadmin3.int.bgdi.ch/mom_kvp_print/index.html?lang=fr&topic=ech&bgLayer=voidLayer&X=247292.01&Y=683252.99&zoom=7&layers=WMTS%7C%7CUebersichtsplan_2011%7C%7Chttp:%2F%2Fwww.gis.stadt-zuerich.ch%2Fwmts%2Fwmts-zh-stzh-ogd.xml)

**External WMTS (KVP): ASIT**

*GetCapbilities* https://ows.asitvd.ch/wmts/1.0.0/WMTSCapabilities.xml 

[Demo](https://mf-geoadmin3.int.bgdi.ch/mom_kvp_print/index.html?lang=fr&topic=ech&bgLayer=ch.swisstopo.swissimage&X=224308.09&Y=605821.62&zoom=8&layers=WMTS%7C%7Casitvd.fond_pourortho%7C%7Chttps:%2F%2Fows.asitvd.ch%2Fwmts%2F1.0.0%2FWMTSCapabilities.xml)

**Timeseries**
Try both single & multi

[Demo](https://mf-chsdi3.int.bgdi.ch/shorten/73e0cd1c2a)

See https://github.com/geoadmin/mf-geoadmin3/issues/3852

### CODE REVIEW

**Comment by davidoesch** [comment]
tested all three links on win10 edge chrome
works like a charm
--prometheus

File: src/components/print/PrintDirective.js [hunk 1]
```diff
@@ -271,44 +271,153 @@ goog.require('ga_urlutils_service');
...
+
         'WMTS': function(layer, config) {
+
+          // config is not defined for external WMTS
+          // For internal WMTS layer, we use the simplified
+          // mapfish print protocol, and the standard for
+          // external WMTS layers.
+          // See http://www.mapfish.org/doc/print/protocol.html#wmts
+
+          var is_external_wmts = angular.equals(config, {});
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: camelCase variable name please

**Comment by gjn** [comment]
Very nice. This is good to merge for me.

@oterral ?

File: src/components/print/PrintDirective.js [hunk 2]
```diff
@@ -272,42 +272,41 @@ goog.require('ga_urlutils_service');
...
+            var resolution = parseFloat(value);
+            var z = tilegrid.getZForResolution(resolution);
+            var tileSize = tilegrid.getTileSize(z);
+            var topLeftCorner = tilegrid.getOrigin(z);
+            var minX = topLeftCorner[0];
+            var maxY = topLeftCorner[1];
+            var maxX = extent[2] || 900000;
+            var minY = extent[1] || 30000;
+            var topLeftTile = tilegrid
+              .getTileCoordForCoordAndZ([minX, maxY], z);
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: bad indentation

File: src/components/print/PrintDirective.js [hunk 3]
```diff
@@ -272,42 +272,41 @@ goog.require('ga_urlutils_service');
...
+          var ids = tilegrid.getMatrixIds();
+          var resolutions = tilegrid.getResolutions();
+          angular.forEach(resolutions, function(value, key) {
+            var resolution = parseFloat(value);
+            var z = tilegrid.getZForResolution(resolution);
+            var tileSize = tilegrid.getTileSize(z);
+            var topLeftCorner = tilegrid.getOrigin(z);
+            var minX = topLeftCorner[0];
+            var maxY = topLeftCorner[1];
+            var maxX = extent[2] || 900000;
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: could we avoid hardcoded value ? You can have this value from the projection extent or the default extent in gaMapUtils.

File: src/components/print/PrintDirective.js [hunk 4]
```diff
@@ -272,42 +272,41 @@ goog.require('ga_urlutils_service');
...
+            var minY = extent[1] || 30000;
+            var topLeftTile = tilegrid
+              .getTileCoordForCoordAndZ([minX, maxY], z);
+            var bottomRightTile = tilegrid
+                .getTileCoordForCoordAndZ([maxX, minY], z);
+            var tileWidth = 1 + bottomRightTile[1] - topLeftTile[1];
+            var tileHeight = 1 + topLeftTile[2] - bottomRightTile[2];
+            //var tileGridSize =  tilegrid.getTileSize(z);
+            //var tileWidth = tileGridSize[0];
+            //var tileHeight = tileGridSize[1];
```
Comment (oterral) [inline src/components/print/PrintDirective.js]: Remove these lines.

File: src/components/map/MapService.js [hunk 5]
```diff
@@ -827,6 +827,7 @@ goog.require('ga_urlutils_service');
                 // Workaround: Set a cache size of zero when layer is
                 // timeEnabled see:
                 // https://github.com/geoadmin/mf-geoadmin3/issues/3491
+                layer: layer.serverLayerName,
```
Comment (oterral) [inline src/components/map/MapService.js]: can you move this line after cacheSize or before the comment please.

**Comment by procrastinatio** [comment]
jenkins tests

**Comment by procrastinatio** [comment]
jenkins tests

**Comment by oterral** [comment]
👍 +1

**Pull request merged**